Mathlib/RingTheory/Kaehler/Localization.lean

```diff
@@ -0,0 +1,143 @@
+`f` is a pre-image of `g` in `R[X]`, this is the `R`-algebra map `R[X,Y] →ₐ[R] (R[X]/I²)[1/f]`
+defined via mapping `Y` to `1/f`. -/
+noncomputable
+def compLocalizationAwayAlgHom : ((Generators.localizationAway g (S := T)).comp P).Ring →ₐ[R]
+      Localization.Away (Ideal.Quotient.mk (P.ker ^ 2) (P.σ g)) :=
+  aeval (R := R) (S₁ := Localization.Away _)
+    (Sum.elim
+      (fun _ ↦ IsLocalization.Away.invSelf <| (Ideal.Quotient.mk (P.ker ^ 2) (P.σ g)))
+      (fun i : P.vars ↦ algebraMap P.Ring _ (X i)))
+
+@[simp]
+lemma compLocalizationAwayAlgHom_toAlgHom_toComp (x : P.Ring) :
+    (compLocalizationAwayAlgHom T g P) (((localizationAway g (S := T)).toComp P).toAlgHom x) =
+      algebraMap P.Ring _ x := by
+  simp only [toComp_toAlgHom, Ideal.mem_comap, RingHom.mem_ker, compLocalizationAwayAlgHom, comp,
+    localizationAway, AlgHom.toRingHom_eq_coe, aeval_rename,
+    Sum.elim_comp_inr, ← IsScalarTower.toAlgHom_apply (R := R), ← comp_aeval_apply,
+    aeval_X_left_apply]
+
+@[simp]
+lemma compLocalizationAwayAlgHom_X_inl : (compLocalizationAwayAlgHom T g P) (X (Sum.inl ())) =
+      IsLocalization.Away.invSelf ((Ideal.Quotient.mk (P.ker ^ 2)) (P.σ g)) := by
+  simp [compLocalizationAwayAlgHom]
+
+lemma compLocalizationAwayAlgHom_relation_eq_zero :
+    (compLocalizationAwayAlgHom T g P) ((rename Sum.inr) (P.σ g) * X (Sum.inl ()) - 1) = 0 := by
+  rw [map_sub, map_one, map_mul, ← toComp_toAlgHom (Generators.localizationAway g (S := T)) P]
+  show (compLocalizationAwayAlgHom T g P) (((localizationAway g).toComp P).toAlgHom _) * _ - _ = _
+  rw [compLocalizationAwayAlgHom_toAlgHom_toComp, compLocalizationAwayAlgHom_X_inl,
+    IsScalarTower.algebraMap_apply P.Ring (P.Ring ⧸ P.ker ^ 2) (Localization.Away _)]
+  simp
+
+lemma sq_ker_comp_le_ker_compLocalizationAwayAlgHom :
+    ((localizationAway g (S := T)).comp P).ker ^ 2 ≤
+      RingHom.ker (compLocalizationAwayAlgHom T g P) := by
+  have hsple {x} (hx : x ∈ Ideal.span {(rename Sum.inr) (P.σ g) * X (Sum.inl ()) - 1}) :
+        (compLocalizationAwayAlgHom T g P) x = 0 := by
+    obtain ⟨a, rfl⟩ := Ideal.mem_span_singleton.mp hx
+    rw [map_mul, compLocalizationAwayAlgHom_relation_eq_zero, zero_mul]
+  rw [comp_localizationAway_ker _ _ (P.σ g) (by simp), sq, Ideal.sup_mul, Ideal.mul_sup,
+    Ideal.mul_sup]
+  apply sup_le
+  · apply sup_le
+    · rw [← Ideal.map_mul, Ideal.map_le_iff_le_comap, ← sq]
+      intro x hx
+      simp only [Ideal.mem_comap, RingHom.mem_ker,
+        compLocalizationAwayAlgHom_toAlgHom_toComp (T := T) g P x]
+      rw [IsScalarTower.algebraMap_apply P.Ring (P.Ring ⧸ P.ker ^ 2) (Localization.Away _),
+        Ideal.Quotient.algebraMap_eq, Ideal.Quotient.eq_zero_iff_mem.mpr hx, map_zero]
+    · rw [Ideal.mul_le]
+      intro x hx y hy
+      simp [hsple hy]
+  · apply sup_le <;>
+    · rw [Ideal.mul_le]
+      intro x hx y hy
+      simp [hsple hx]
+
+/-- `T ⊗[S] (I/I²) → J/J²` is injective if `T` is the localization of `S` away from an element. -/
```

Comment:
I don't think `J` is defined anywhere else? And I think we should also copy the definition of `I` (or the whole setting) here.